File: 2-AD-HOC/1588-Ajude_a_federacao-7/1588.py
for i in range(int(input())):
    times={}
    info=input().split(' ')
    for g in range(int(info[0])):
        times[input()]=[0,0,0]
    for g in range(int(info[1])):
        jogo=input().split(' ')
        if int(jogo[0]) > int(jogo[2]):
            times[jogo[1]][0]+=3
            times[jogo[1]][1]+=1
        elif int(jogo[0]) < int(jogo[2]):
            times[jogo[3]][0]+=3
            times[jogo[3]][1]+=1
            
        else:
            times[jogo[3]][0]+=1
            times[jogo[1]][0]+=1
                
            
        times[jogo[1]][2]+=1*int(jogo[0])
        times[jogo[3]][2]+=1*int(jogo[2])
        
        times_ordenados = sorted(times.items(), key=lambda item: (-item[1][0], -item[1][1], -item[1][2]))
    for a in times_ordenados:
        print(a[0])


            

File: 2-AD-HOC/1588-Ajude_a_federacao-7/1588.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    static void Main()
    {
        int n = int.Parse(Console.ReadLine());
        for (int i = 0; i < n; i++)
        {
            var times = new Dictionary<string, int[]>();
            var info = Console.ReadLine().Split(' ');

            int numTimes = int.Parse(info[0]);
            int numJogos = int.Parse(info[1]);

            for (int g = 0; g < numTimes; g++)
            {
                string time = Console.ReadLine();
                times[time] = new int[] { 0, 0, 0 };
            }

            for (int g = 0; g < numJogos; g++)
            {
                var jogo = Console.ReadLine().Split(' ');

                int golTime1 = int.Parse(jogo[0]);
                int golTime2 = int.Parse(jogo[2]);
                string time1 = jogo[1];
                string time2 = jogo[3];

                if (golTime1 > golTime2)
                {
                    times[time1][0] += 3;
                    times[time1][1] += 1;
                }
                else if (golTime1 < golTime2)
                {
                    times[time2][0] += 3;
                    times[time2][1] += 1;
                }
                else
                {
                    times[time1][0] += 1;
                    times[time2][0] += 1;
                }

                times[time1][2] += golTime1;
                times[time2][2] += golTime2;
            }

            var timesOrdenados = times.OrderByDescending(t => t.Value[0])
                                      .ThenByDescending(t => t.Value[1])
                                      .ThenByDescending(t => t.Value[2]);

            foreach (var time in timesOrdenados)
            {
                Console.WriteLine(time.Key);
            }
        }
    }
}